### PR TITLE
search-engines: Adds new engines and tweaks define-search-engine

### PR DESCRIPTION
**Pull request by ghost**
This adds a few more search engines. Some things to consider:

- I tweaked `define-search-engine` as per our discussion on #7, giving the ability for delimiters to be written by hand if the search engine has a weird query syntax (as is the case with Lemmy, in this case).
- Also in the case of Lemmy, its search engine needs all of its query parameters to be manually supplied in order for it to work, otherwise it returns a server error, so I added another macro argument for this.
- I added Whoogle and Teddit engines as derived engines from Google and Reddit respectively. Are alternative front-ends worthy of a search engine? Should they inherit from their original engine?
- In regard to the above, if a derived search engine is defined, it currently can't take a `:documentation` slot, should this be the case?
- Are completion functions such as `make-invidious-completion` to depend on the engine's `base-search-url` for their own `base-search-url`? I believe the current behavior is inconsistent (e.g. search results might work but completion may not if the completion URL is down), specially for self-hostable search engines, and this would also allow derived engines to use their base search url for completion instead of the original engine's.

### CODE REVIEW

**Comment by aartaka** [comment]
> * I tweaked `define-search-engine` as per our discussion on [search-engines.lisp: Adds peertube, sourcehut, and libgen engines #7](https://github.com/aartaka/nx-search-engines/pull/7), giving the ability for delimiters to be written by hand if the search engine has a weird query syntax (as is the case with Lemmy, in this case).

While I hope that there's not much engines with weird syntax, I am most probably mistaken :) Thanks for this addition!

> * Also in the case of Lemmy, its search engine needs all of its query parameters to be manually supplied in order for it to work, otherwise it returns a server error, so I added another macro argument for this.

Hmmmm. Why put it in the README then? I mean, we either fix the URL automatically or make the user of `nx-search-engines` to do it. In this case, it's either a macro argument or documentation, not both.

> * I added Whoogle and Teddit engines as derived engines from Google and Reddit respectively. Are alternative front-ends worthy of a search engine? Should they inherit from their original engine?

Free instances are definitely worthy of a separate function -- for the sake of better representation, at least. If the query syntax is the same, then yes -- they should be derived.

> * In regard to the above, if a derived search engine is defined, it currently can't take a `:documentation` slot, should this be the case?

Ha, that I didn't consider! That's a bug in `define-derived-search-engine`! Are you interested in trying to fix it, or should I do it?

> * Are completion functions such as `make-invidious-completion` to depend on the engine's `base-search-url` for their own `base-search-url`? I believe the current behavior is inconsistent (e.g. search results might work but completion may not if the completion URL is down), specially for self-hostable search engines, and this would also allow derived engines to use their base search url for completion instead of the original engine's.

Completion URL is hard-coded into the completion function, and there's no way to redefine completion URL without redefining the completion function. This is both useful (to make absolutely insane completion functions enhancing original engines) and inconvenient (like in your case with hostable search engines). I'm not sure what to do here.

File: search-engines.lisp [hunk 1]
```diff
@@ -1202,6 +1219,115 @@ one of these at once unlike on the web interface.")
...
+             (:week "week")
+             (:day "day")
+             (:hour "hour")))
+  (type-from "type" ((:link "link")
+                     (:comments "comment")
+                     (:communities "sr")
+                     (:user "user"))))
+
+(define-derived-search-engine teddit
+    (:shortcut "teddit"
```
Comment (aartaka) [inline search-engines.lisp]: `reddit` forgotten, as the parent engine, first in the parenthesized list.

File: README.org [hunk 2]
```diff
@@ -561,6 +562,52 @@ Search engine for [[https://libgen.li/][Library Genesis]], an aggregator of text
...
+       [:page integer])
+
+#+end_src
+
+** Invidious
+Search engine for [[https://invidious.io/][Invidious]], an open source alternative front-end to YouTube.
+
+#+begin_src
+(invidious [:shortcut string]
+           [:fallback-url quri:uri]
```
Comment (aartaka) [inline README.org:653]: Maybe we should mention the `:completion-function` as a valid argument in the documentation, given the problem with hard-coded completion URLs?

File: README.org [hunk 3]
```diff
@@ -561,6 +562,52 @@ Search engine for [[https://libgen.li/][Library Genesis]], an aggregator of text
...
+           [:base-search-url string]
+           [:upload-date {:none | :hour | :today | :week | :month | :year }]
+           [:result-type {:all | :video | :channel | :playlist | :movie | :show }]
+           [:duration {:none | :short | :long | :medium}]
+           [:features {:none | :live | :4k | :hd | :subtitles | :cc | :vr-180 |
+                       :3d | :hdr | :location | :purchased}]
+           [:sort-by {:relevance | :rating | :date | :views}])
+#+end_src
```
Comment (aartaka) [inline README.org:662]: Note that if an engine has a completion function constructor (e.g. `make-invidious-completion-function`), it should be documented too :)

**Review by aartaka — CHANGES_REQUESTED** [review]

**Comment by aartaka** [comment]
@efimerspan, friendly ping :) If you're bored, I can merge and finish it myself.

**Comment by ghost** [comment]
Hey @aartaka, apologies, I wasn't able to get around to doing this. I'll try to submit my changes along with a couple of more search engines by the end of the week if this is not an issue to you.

**Comment by aartaka** [comment]
Sure thing, I'm in no hurry :)

**Comment by aartaka** [comment]
To get back to the completion function question: we can add a `url` argument to a completion function, so that it's easier to change. E.g.
``` lisp
(defun make-invidious-completion (url &key request-args) ...)
```

**Comment by ghost** [comment]
@aartaka Apologies for the delay.

> To get back to the completion function question: we can add a `url` argument to a completion function, so that it's easier to change. E.g.
> 
> ```lisp
> (defun make-invidious-completion (url &key request-args) ...)
> ```

Addressed with an `instance` argument.

> Hmmmm. Why put it in the README then? I mean, we either fix the URL automatically or make the user of nx-search-engines to do it. In this case, it's either a macro argument or documentation, not both.

I left it as a macro argument, otherwise it will be a pain for the user to input all of the filters IMO.

> Ha, that I didn't consider! That's a bug in define-derived-search-engine! Are you interested in trying to fix it, or should I do it?

I tried to implement a workaround for this without affecting existing engines, please let me know if there's a better way to do so.

> Free instances are definitely worthy of a separate function -- for the sake of better representation, at least. If the query syntax is the same, then yes -- they should be derived.

In regard to this, I changed the order of the arguments in the `define-derived-search-engine`, as currently, if the user tries to override a derived engine's `base-search-url`, for instance, it won't have an effect. Now, what should we do in the case of a derived engine such as Whoogle, which takes all of Google's search parameters *plus* its own search parameters?

**Comment by aartaka** [comment]
> Addressed with an `instance` argument.

Thanks!

> I left it as a macro argument, otherwise it will be a pain for the user to input all of the filters IMO.

Yes, that's most reasonable.

> I tried to implement a workaround for this without affecting existing engines, please let me know if there's a better way to do so.

No, it's just fine. I thought about this docstring injection myself too!

> > Free instances are definitely worthy of a separate function -- for the sake of better representation, at least. If the query syntax is the same, then yes -- they should be derived.
> 
> In regard to this, I changed the order of the arguments in the `define-derived-search-engine`, as currently, if the user tries to override a derived engine's `base-search-url`, for instance, it won't have an effect.

Good catch!

> Now, what should we do in the case of a derived engine such as Whoogle, which takes all of Google's search parameters _plus_ its own search parameters?

Well, in this case we can either 
- Re-implement part of the logic from `define-search-engine` in `define-derived-search-engine`, or
- Re-implement part of the keys from `google` in `whoogle`, with more on top.

I'm fine with either of the two, as much as I'm fine with having this PR merged as it is. What do you think?

**Comment by ghost** [comment]
> Well, in this case we can either
> - Re-implement part of the logic from define-search-engine in define-derived-search-engine, or
> - Re-implement part of the keys from google in whoogle, with more on top.
I'm fine with either of the two, as much as I'm fine with having this PR merged as it is. 

> What do you think?

I went with the second approach, mainly because some of `google` engine's keys have a `whoogle` counterpart as per [this file](https://github.com/benbusby/whoogle-search/blob/194ddc33f36e6955587816acd2347bbebeef7912/app/request.py#L20) and this [other file](https://github.com/benbusby/whoogle-search/blob/194ddc33f36e6955587816acd2347bbebeef7912/app/models/config.py#L34). I've documented this discrepancy. 

Given as I was adding the corresponding `whoogle` keywords, I noticed the `google` engine was quite barebones, so I introduced some of the keywords that made sense to me to have in this engine, and left out other such as `adtest`, `source`, `btnI` (feeling lucky), and similar which don't really add much to the functionality of the engine. I also noticed that the list of countries and languages that both `whoogle` and `google` engines take for some filters is quite extensive, so instead of duplicating it I added some helper functions and leveraged the `:function` value.

On a different note, one thing I've seen when adding some engines, is that in some cases (e.g. `extra-filters` on `google` or the various column filters on `libgen`), URL parameters can take a comma-separated list of the provided options, maybe we could let the user supply a keyword list in this case?

Also, I've just noticed the new contributing guidelines. I'll ensure to split out my contributions into separate PRs from now onward to make it easier for you to review them.

File: search-engines.lisp [hunk 4]
```diff
@@ -1202,6 +1648,233 @@ one of these at once unlike on the web interface.")
...
+           :fallback-url (quri:uri "https://codeberg.org/explore/organizations"))
+    "`gitea' derived search engine for a Gitea's instance organizations.")
+
+(define-search-engine hacker-news
+    (:shortcut "hacker-news"
+     :fallback-url (quri:uri "https://hn.algolia.com")
+     :base-search-url "https://hn.algolia.com/?q=~a"
+     :documentation "`nyxt:search-engine' for Hacker News via Algolia Search.")
+  (date-range "dateRange" ((:all "all")
+                           (:past-24-hours "last24h")
```
Comment (aartaka) [inline search-engines.lisp]: Maybe have `:past-day` instead/in addition to `:past-24-hours` for consistency?

**Comment by aartaka** [comment]
Okay, it looks just fine now, I am more than ready to merge :)

> I went with the second approach, mainly because some of `google` engine's keys have a `whoogle` counterpart as per [this file](https://github.com/benbusby/whoogle-search/blob/194ddc33f36e6955587816acd2347bbebeef7912/app/request.py#L20) and this [other file](https://github.com/benbusby/whoogle-search/blob/194ddc33f36e6955587816acd2347bbebeef7912/app/models/config.py#L34). I've documented this discrepancy.

Sure, thanks!

> Given as I was adding the corresponding `whoogle` keywords, I noticed the `google` engine was quite barebones, so I introduced some of the keywords that made sense to me to have in this engine, and left out other such as `adtest`, `source`, `btnI` (feeling lucky), and similar which don't really add much to the functionality of the engine. I also noticed that the list of countries and languages that both `whoogle` and `google` engines take for some filters is quite extensive, so instead of duplicating it I added some helper functions and leveraged the `:function` value.

Indeed, `google` engine was quite forgotten :D Thanks for taking care of it!

And yes, `:function` is your friend in all the compicated cases like this.

> On a different note, one thing I've seen when adding some engines, is that in some cases (e.g. `extra-filters` on `google` or the various column filters on `libgen`), URL parameters can take a comma-separated list of the provided options, maybe we could let the user supply a keyword list in this case?

Yes, and process it with `:function`!

> Also, I've just noticed the new contributing guidelines. I'll ensure to split out my contributions into separate PRs from now onward to make it easier for you to review them.

Thanks for that. I hope that it won't be too burdensome for you ❤️

**Comment by aartaka** [comment]
So, do I merge this now, or is there something else left to do here, @efimerspan?

**Comment by ghost** [comment]
@aartaka That is all, thank you for the patience :)

**Comment by aartaka** [comment]
Thank you for this huge contribution too!

**Pull request merged**